### PR TITLE
clickhouse connector

### PR DESCRIPTION
**Pull request by ellison840611**
== RELEASE NOTES ==

General Changes
* Developed a new presto connector for clickhouse

### CODE REVIEW

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        this.passwordCredentialName = passwordCredentialName;
+        return this;
+    }
+
+    public boolean isCaseInsensitiveNameMatching()
+    {
+        return caseInsensitiveNameMatching;
+    }
+
+    @Config("clickhouse.case-insensitive-name-matching")
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java:110]: clickhouse.case-sensitive?

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    @NotNull
+    @MinDuration("0ms")
+    public Duration getCaseInsensitiveNameMatchingCacheTtl()
+    {
+        return caseInsensitiveNameMatchingCacheTtl;
+    }
+
+    @Config("clickhouse.case-insensitive-name-matching.cache-ttl")
+    public ClickHouseConfig setCaseInsensitiveNameMatchingCacheTtl(Duration caseInsensitiveNameMatchingCacheTtl)
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java:125]: clickhouse.remote-name-cache-ttl

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConnectorFactory.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public Connector create(String catalogName, Map<String, String> requiredConfig, ConnectorContext context)
+    {
+        requireNonNull(requiredConfig, "requiredConfig is null");
+
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            Bootstrap app = new Bootstrap(
+                    binder -> {
+                        //binder.bind(TypeManager.class).toInstance(context.getTypeManager());
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConnectorFactory.java:67]: let's remove this line

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseEngineType.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.clickhouse;
+
+public enum ClickHouseEngineType
+{
+    STRIPELOG("StripeLog"),
+    LOG("Log"),
+    TINYLOG("TinyLog"),
+    MERGETREE("MergeTree()"),
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseEngineType.java:21]: why MergeTree has `()`? could we import clickhouse engine type from clickhouse classes?

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/DriverConnectionFactory.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class DriverConnectionFactory
+        implements ConnectionFactory
+{
+    private final ClickHouseDriver driver;
+    private final String connectionUrl;
+    private final Properties connectionProperties;
+    private final Optional<String> userCredentialName;
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/DriverConnectionFactory.java:33]: s/userCredentialName/userCredential/g

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/DriverConnectionFactory.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    private final Optional<String> passwordCredentialName;
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/DriverConnectionFactory.java:34]: s/passwordCredentialName/passwordCredential/g

File: presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseConfig.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+                .setAllowDropTable(false)
+                .setCaseInsensitiveNameMatchingCacheTtl(new Duration(1, MINUTES))
+                .setMapStringAsVarchar(false));
+    }
+
+    @Test
+    public void testExplicitPropertyMappings()
+    {
+        Map<String, String> properties = new ImmutableMap.Builder<String, String>()
+                .put("clickhouse.connection-url", "jdbc:h2:mem:config")
```
Comment (zhenxiao) [inline presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseConfig.java:47]: let's fix all these properties names

File: presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseDistributedQueries.java [hunk 8]
```diff
@@ -0,0 +1,782 @@
...
+    }
+
+    @Test
+    @Override
+    public void testInsertIntoNotNullColumn()
+    {
+        skipTestUnless(supportsNotNullColumns());
+
+        String catalog = getSession().getCatalog().get();
+//        String createTableStatement = "CREATE TABLE " + catalog + ".tpch.test_not_null_with_insert (\n" +
```
Comment (zhenxiao) [inline presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseDistributedQueries.java:185]: let's remove all these commented code

File: presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseDistributedQueries.java [hunk 9]
```diff
@@ -0,0 +1,782 @@
...
+                "\"column_a\"",
+                "\"column_b\"",
+                "\"column_c\"");
+        assertUpdate(createTableSql);
+        assertEquals(computeScalar("SHOW CREATE TABLE test_not_null_with_insert"), expectedCreateTableSql);
+
+        assertQueryFails("INSERT INTO test_not_null_with_insert (column_a) VALUES (date '2012-12-31')", "(?s).*NULL.*column_b.*");
+        assertQueryFails("INSERT INTO test_not_null_with_insert (column_a, column_b) VALUES (date '2012-12-31', null)", "(?s).*NULL.*column_b.*");
+
+        //assertUpdate("ALTER TABLE test_not_null_with_insert ADD COLUMN column_c BIGINT NOT NULL");
```
Comment (zhenxiao) [inline presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseDistributedQueries.java:234]: remove commented code

File: presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseDistributedQueries.java [hunk 10]
```diff
@@ -0,0 +1,782 @@
...
+        assertUpdate("INSERT INTO test_not_null_with_insert (column_b, column_c) VALUES (date '2012-12-31', 1)", 1);
+        assertUpdate("INSERT INTO test_not_null_with_insert (column_a, column_b, column_c) VALUES (date '2013-01-01', date '2013-01-02', 2)", 1);
+        assertQuery(
+                "SELECT * FROM test_not_null_with_insert",
+                "VALUES ( NULL, CAST ('2012-12-31' AS DATE), 1 ), ( CAST ('2013-01-01' AS DATE), CAST ('2013-01-02' AS DATE), 2 );");
+
+        assertUpdate("DROP TABLE test_not_null_with_insert");
+    }
+
+//
```
Comment (zhenxiao) [inline presto-clickhouse/src/test/java/com/facebook/presto/plugin/clickhouse/TestClickHouseDistributedQueries.java:255]: remove commented code

File: presto-docs/src/main/sphinx/connector/clickhouse.rst [hunk 11]
```diff
@@ -0,0 +1,156 @@
+====================
+ClickHouse connector
+====================
+
+The ClickHouse connector allows querying tables in an external
+`Yandex ClickHouse <https://clickhouse.tech/>`_ server. This can be used to
```
Comment (zhenxiao) [inline presto-docs/src/main/sphinx/connector/clickhouse.rst:6]: let's remove Yandex

File: presto-docs/src/main/sphinx/connector/clickhouse.rst [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+query data in the databases on that server, or combine it with other data
+from different catalogs accessing ClickHouse or any other supported data source.
+
+Requirements
+------------
+
+To connect to a ClickHouse server, you need:
+
+* ClickHouse version 20.8 or higher.
+* Network access from the Trino coordinator and workers to the ClickHouse
```
Comment (zhenxiao) [inline presto-docs/src/main/sphinx/connector/clickhouse.rst:16]: s/Trino/Presto/g

**Review by zhenxiao — CHANGES_REQUESTED** [review]
looks nice, @ellison840611 
a few more comments

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseDXLKeyWords.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.clickhouse;
+
+public class ClickhouseDXLKeyWords
```
[the remark this answers is not in the document]
Comment (ellison840611) [reply presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseDXLKeyWords.java:16]: thanks a lot for your review. On this issue, the information currently known is：clickhouse is written in c++, and no definition of such constants is found in the jdbc jar of clickhouse.

[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickhouseDXLKeyWords.java:16]: get it. In this case, no need to import

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    @Nullable
+    public String getPasswordCredentialName()
+    {
+        return passwordCredentialName;
+    }
+
+    @Config("clickhouse.password-credential")
+    public ClickHouseConfig setPasswordCredentialName(String passwordCredentialName)
```
Comment (zhenxiao) [inline presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java:100]: s/setPasswordCredentialName/setPasswordCredential/g

File: presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+        return this;
+    }
+
+    @Nullable
+    public String getUserCredentialName()
+    {
+        return userCredentialName;
+    }
+
+    @Config("clickhouse.user-credential-name")
```
[the remark this answers is not in the document]
Comment (zhenxiao) [reply presto-clickhouse/src/main/java/com/facebook/presto/plugin/clickhouse/ClickHouseConfig.java:86]: s/setUserCredentialName/setUserCredential/g

**Review by zhenxiao — APPROVED** [review]
@ellison840611 mostly good
a few minor things

**Comment by zhenxiao** [comment]
hi @beinan @highker could you please take a look at Clickhouse connector? This PR looks good to me

**Review by highker — COMMENTED** [review]
@zhenxiao, I will pass the review given the connector change is pretty much contained. Feel free to merge if it looks good to you.

**Pull request merged**